### DISCUSSION TITLE
Support for EIP712 signers in ZkSync

### DISCUSSION THREAD

**Discussion by mariopil** [wevm/wagmi#3218] [Ideas]
Hello, I'm currently developing support for EIP712 signers in ZkSync in `viem` (PR opened [here](https://github.com/wagmi-dev/viem/pull/1461) ) and have recently started testing the implementation with `wagmi`. From what I have noticed there are small changes required in `wagmi` to be able to properly send EIP712 transactions which contain few additional fields. Hooks like `useSendTransaction` or `useContractWrite` have all transaction fields specified explicitly, therefore all the EIP712 fields would need to be added there too (and also in all methods that are called from there internally). 

Currently the transaction structure in `viem` is flat, as discussed [here](https://github.com/wagmi-dev/viem/discussions/1229) so that would mean adding 5 new fields everywhere:
```
gasPerPubdata
paymaster
factoryDeps
paymasterInput
customSignature
```
The other approach could be to put all those fields into separate structure, like it was initially proposed, and then add only `customData` param in `wagmi` hooks:
```
customData: {
    gasPerPubdata
    factoryDeps
    customSignature
    paymasterParams: {
        paymaster
        paymasterInput
    }
}
```
That would require some changes in the `viem` PR too. 

Before I start implementing changes I would like to have your opinion about it. What do you think would be the best approach here?

**Comment by RobbyUitbeijerse** [comment]
hey @mariopil, I'm currently working on something similar, and noticed that this discussion didn't actually proceed since 2023. Did you figure out the wagmi implementation on your end?

We've been thinking about handling it by hijacking each request to the RPC through passing `onFetchRequest` to the http() transport, essentially doing something like the snippet below. Would love to know about insights you may have come across since last year. 

```
  http(undefined, {
    onFetchRequest(request) {
      if (request.method === 'eth_sendTransaction') {

        const body = JSON.parse(request.body);

        body.params[0].paymaster = '...';
        ...

        request.body = JSON.stringify(body);
      }

      return request;
    },
  })
```

**Reply by mariopil** [reply]
Hi, I've actually ended up without changing anything on the `wagmi` side - everything was done in [`viem` only](https://github.com/wevm/viem/pull/1665/files#diff-5dc7858aa97b9f146d18d05c455c0f0fa7fad9692b765749528870a66bb77478)
